cirq-google/cirq_google/engine/stream_response_demux.py

```diff
@@ -0,0 +1,94 @@
+from collections.abc import Callable
+
+import asyncio
+import duet
+
+from cirq_google.cloud import quantum
+
+JobPath = str
+MessageId = str
+CancelCallback = Callable[[quantum.QuantumRunStreamRequest], None]
+
+
+class _StreamResponseDemux:
+    """A event demultiplexer for QuantumRunStreamResponses, as part of the async reactor pattern.
+
+    Args:
+        cancel_callback: Function to be called when the future matching its request argument is
+        canceled.
+    TODO(#5996) expand
+    """
+
+    def __init__(self, cancel_callback: CancelCallback):
+        self._subscribers: dict[JobPath, dict[MessageId, duet.AwaitableFuture]] = {}
+        self._cancel_callback = cancel_callback
+
+    def subscribe(
+        self, request: quantum.QuantumRunStreamRequest
+    ) -> duet.AwaitableFuture[quantum.QuantumRunStreamResponse]:
+        """Subscribes to the QuantumRunStreamResponse matching the given request.
+
+        The request must be unique among all calls to this method, with message ID as the
+        identifier.
+
+        Returns:
+            A future for the response to be fulfilled when publish is called.
+        Raises:
+            ValueError: If there is already a subscriber for the given request.
+        """
+
+        request_type = request.WhichOneOf('request')
+        if request_type == 'create_quantum_program_and_job':
+            job_path = request.create_quantum_program_and_job.quantum_job.name
+        elif request_type == 'create_quantum_job':
+            job_path = request.create_quantum_job.quantum_job.name
+        else:  # request_type == 'get_quantum_result'
+            job_path = request.get_quantum_result.parent
+
+        if job_path not in self._subscribers:
+            self._subscribers[job_path] = {}
+
+        subscribers_by_message = self._subscribers[job_path]
+        if request.message_id in subscribers_by_message:
+            raise ValueError(f'There is another subscriber for the message ID {request.message_id}')
+
+        def cancel(future: asyncio.Future):
```

Comment:
Note, the future passed in here will be the `response_future` created below, which is what we're adding this callback to, so it will have type `duet.AwaitableFuture[quantum.QuantumRunStreamResponse]`.

Reply:
Thanks for the catch! Meant to have the superclass of `AwaitableFuture` here, but I think having `AwaitableFuture` makes more sense since that's the only possible type.